idea/idea-fir/src/org/jetbrains/kotlin/idea/core/overrideImplement/KtGenerateMembersHandler.kt

```diff
@@ -0,0 +1,226 @@
+            if (firstExistingEntry != null) {
+                firstAnchor = firstExistingEntry.psi.prevSiblingOfSameType() ?: classLeftBrace
+            }
+        }
+
+        val entryBatchAndAnchor = mutableListOf<Pair<List<KtCallableDeclaration>, PsiElement?>>()
+        var currentAnchor = firstAnchor
+        val currentBatch = mutableListOf<KtCallableDeclaration>()
+        fun updateBatch() {
+            if (currentBatch.isNotEmpty()) {
+                entryBatchAndAnchor += currentBatch.toList() to currentAnchor
+                currentBatch.clear()
+            }
+        }
+        for (entry in symbolsInPreferredOrder) {
+            when (entry) {
+                is MemberEntry.ExistingEntry -> {
+                    updateBatch()
+                    currentAnchor = entry.psi
+                }
+                is MemberEntry.NewEntry -> {
+                    currentBatch += entry.psi
+                }
+            }
+        }
+        updateBatch()
+
+        var firstElement: PsiElement? = null
+        for ((newDeclarations, anchor) in entryBatchAndAnchor) {
+            // TODO: implement proper referenceShortener
+            val inserted = insertMembersAfter(editor, currentClass, newDeclarations, anchor = anchor, referenceShortener = {})
+            if (firstElement == null) {
+                firstElement = inserted.first()
+            }
+        }
+
+        firstElement?.let {
+            moveCaretIntoGeneratedElement(editor, it)
+        }
+    }
+
+    private class DoublyLinkedNode<T>(val t: T? = null, var prev: DoublyLinkedNode<T>? = null, var next: DoublyLinkedNode<T>? = null) {
+        fun append(node: DoublyLinkedNode<T>): DoublyLinkedNode<T> {
+            val next = this.next
+            this.next = node
+            node.prev = this
+            node.next = next
+            next?.prev = node
+            return node
+        }
+
+        fun prepend(node: DoublyLinkedNode<T>): DoublyLinkedNode<T> {
+            val prev = this.prev
+            this.prev = node
+            node.next = this
+            node.prev = prev
+            prev?.next = node
+            return node
+        }
+
+        @OptIn(ExperimentalStdlibApi::class)
+        fun toListSkippingNulls(): List<T> {
+            var current: DoublyLinkedNode<T>? = this
+            return buildList {
+                while (current != null) {
+                    current?.let {
+                        if (it.t != null) add(it.t)
+                        current = it.next
+                    }
+                }
+            }
+        }
+    }
+
+    private sealed class MemberEntry {
+        data class ExistingEntry(val psi: KtCallableDeclaration) : MemberEntry()
+        data class NewEntry(val psi: KtCallableDeclaration) : MemberEntry()
+    }
+
+    companion object {
+        @JvmStatic
+        protected val renderOption = KtDeclarationRendererOptions(
```

Comment:
Why is it JvmStatic?

Reply:
It's to make it work with `protected`. But now since everything is `internal`. I don't think we still need `protected`. Let me remove both.